pkg/clockcache/cache.go

```diff
@@ -0,0 +1,201 @@
+package clockcache
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+// CLOCK based approximate LRU storing mappings from strings to
+// int64s designed for concurrent usage.
+// It is sharded into 127 blocks, each of which can be
+// locked independently. Gets only require a read lock on the
+// individual block the element can be found in. When the cache is
+// not full, Insert only requires a write lock on the individual
+// block. Eviction locks blocks one at a time looking for a value
+// that's valid to evict/
+type Cache struct {
+	lock sync.RWMutex
+	// stores indexes into storage
+	elements map[interface{}]*Element
+	storage  []Element
+
+	// CLOCK sweep state, must have write lock
+	next int
+}
+
+type Element struct {
+	// The value stored with this element.
+	key   interface{}
```

Comment:
Yeah, I don't love this. The problem we have is that we want to use both `int64`s and `string`s as keys, and I'm not sure @cevian is willing for us to `generate` a unique cache type for each. I considered storing redundant key and fields for different types, but since we'd need to `interface{}` the keys anyway to store in the `map`, I don't think it's worthwhile.
(`interface{}`ing will _also_ increase the cache miss rate normally, because of the pointer-chasing you mentioned, the padding below is to prevent false-sharing and bouncing elements around cores because of new adjacent elements. I'm not sure it's worthwhile in this case, because we should be read-mostly anyway, and I have concerns about the `map`, but ehh, unless we have more stuff to store here, it doesn't really matter)